### DISCUSSION TITLE
@NativeImageTest and Integration Test with @Inject. Any solution?

### DISCUSSION THREAD

**Discussion by heliommsfilho** [quarkusio/quarkus#24775] [Q&A]
Nowadays there is some way to run a `@NativeImageTest` extending from a `QuarkusIntegrationTest` with `@Inject`?

I have an integration test that injects a `PgPool` but the test fails when building the binary:  
```
org.junit.platform.commons.JUnitException: @Inject is not supported in @NativeImageTest and @QuarkusIntegrationTest tests. Offending field is...
```

If not, how do you manage this situation when we need a real bean to perform a real operation?

**Comment by t1** [comment]
`@Inject` is intended for unit tests (see [here](https://quarkus.io/guides/getting-started-testing#injection-into-tests)), i.e. for tests running within a CDI container containing all the beans of your application... all is within one JVM. When running integration tests, there are two JVMs, one for your application and one for your tests; i.e. you can't inject beans from one JVM into the other.

It would be possible for Quarkus to just build a new CDI context for the integration tests, and build all necessary beans a second time in the test JVM. This would be handy for some cases, like yours. But it would also be very confusing in many other cases, as then there would be in fact two instances; and a lot of unit tests that use `@Inject` are not ready for this and would fail without an easy explanation.

You'll have to find a different method to create your `PgPool`. And as the `@Inject` is still in your (unit) test class, you'll have to change the inheritance structure. I like to make the `SomeTest` class abstract and push the `@Inject` down into a new `SomeUnitTest` class. Then I write a getter for that field and define it in the `SomeTest` class as `abstract`... and use it in the tests. Finally, I have to override that getter in `SomeIT`.

BTW: Maybe you should find a better abstraction than using `PgPool` directly.
